server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java

```diff
@@ -68,18 +95,25 @@ public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation alloca
      */
     @Override
     public Decision canRebalance(RoutingAllocation allocation) {
```

Comment:
As discussed, I think we want this to be 

```java
Decision canRebalance(allocation) {
   return frozenRebabalanceLimitExceeded(allocation) && regularRebalanceExceeded(allocation) ? THROTTLE : YES
}
```

Reply:
Could also just return `YES` here and throttle in the shard-specific method, but checking limits here gives an opportunity for early exit.